### ISSUE TITLE
Routes with query params in Angular 6 not working on page reload

### ISSUE THREAD

**Issue by ankitko** [angular/angular#30329]
Currently, when we are routing to any route in Angular 6 with query params 
like this.router.navigate(['studentInformation'], {queryParams: {filter: '12345'}});

I am able to route and url becomes
http://localhost:4200/#/studentInformation?filter=12345

but when we reload the page on that same URL, We are getting the error 

Error: Cannot match any routes. URL Segment: 'studentInformation%3Ffilter%3D12345'

I tried to remove character encoding in URl by implementing CustomUrlSerializer like

export class CustomUrlSerializer implements UrlSerializer {

    parse(url: string): UrlTree {
        // Change plus signs to encoded spaces
        url.replace("%3F", '?');
        // Use the default serializer that you can import to just do the 
        // default parsing now that you have fixed the url.
        return url  
    }

    serialize(tree: UrlTree): string {
        // Use the default serializer to create a url and replace any spaces with + signs
        return super.serialize(tree).replace("%3F", '?');
    }
}

after implementing this encoded character got removed from URL on reload but still getting error from the angular router
 
Error: Cannot match any routes. URL Segment: 'studentInformation%3Ffilter%3D12345'

I think angular router is trying to match full url including the query params and hence it is failing.
Ideally Angular router should ignore the query params while matching for the routes on reload.
Is this a bug in Angular? Any workaround for this?

**Comment by jasonaden** [comment]
I'm sorry but reported issues require a reproduction of the issue. We suggest using [StackBlitz](https://stackblitz.com/fork/angular-issue-repro2).

If this issue persists, please create a reproduction using one of the links above. Describe the difference between the expected and current behavior in a new issue with the reproduction linked.

<br>

I suggest trying to update to a newer version. But without a reproduction, we can't look into this issue.

**Comment by byvlstr** [comment]
@ankitko How did you solve your problem in the end? I am having a similar problem but on Angular 7.0.

**Comment by ankitko** [comment]
@jasonaden I was not able to solve this problem. Looks like this is an issue in Angular 6 and may be in the version 7 as well. Are you facing this issue on child routes or even on parent routes reloads?

**Comment by byvlstr** [comment]
@ankitko on parent routes with query params

**Comment by Anar21** [comment]
@ankitko @byvlstr Guys I was having exactly the same problem with yours and I was able to solve it. Hope this solution is going to work with you.

Problem is: when you reloading the page with the same **query** parameters angular just append it to route **parameter**. 

What do you have to do is in the function that you are going to reload or navigate you just need to remove the query parameter first. So you will get a newUrl with only parameters not query parameters .
Only after that in the router.navigate you will add your newUrl and add your query parameter. 

Let me show you some example:

So here in the setResource function, I change my parameter but my query parameter stays the same. 
So I get my new parameter then I replace my whole URL and after that, I remove my query parameter using string.slice() method and only after that I use the router.navigate and as a parameter, I put my clean URL without any query parameter and as extras, I add query parameters. It should work

private setResource(selectedResource: Resource)
  {
    this.updatedResource = selectedResource;
    const url = this.router.url.replace(this.resource.id, this.updatedResource.id);
    const index = url.indexOf('?');
    const newUrl = index > -1 ? url.slice(0, index) : url;
    this.router.navigate([newUrl], { queryParams:{student: 'Jane'} });
    this.resource = selectedResource;
  }

**Comment by byvlstr** [comment]
@Anar21 That's interesting but I feel like my issue is slightly different. In fact, I am able to access the page, say "/user?gender=m" when I am already on the app, on any of the pages. The problem only occurs when I either refresh the page "/user?gender=m" or when I use a bookmark to this URL  (from a different website). 

I suppose something must be correct with the routing, but at the same time, some part obviously fails. Fyi, we use lazy loading on our modules, someone suggested this might relate to the issue.
